lib/DiscoveryService.php

```diff
@@ -0,0 +1,192 @@
+
+	public function __construct(
+		AppConfig $config,
+		ILogger $logger,
+		ICacheFactory $cacheFactory,
+		IClientService $httpClient
+	) {
+		$this->appConfig = $config;
+		$this->logger = $logger;
+		$this->cache = $cacheFactory->create('oca.richdocuments.discovery');
+		$this->httpClient = $httpClient;
+	}
+
+	/**
+	 * Return the original wopi url or test wopi url
+	 */
+	public function getWopiUrl(): string {
+		if ($this->appConfig->testUserSessionEnabled()) {
+			return $this->appConfig->getAppValue('test_wopi_url');
+		}
+		return $this->appConfig->getAppValue('wopi_url');
+	}
+
+	/**
+	 * Get urlsrc and action for a given mimetype from WOPI discovery
+	 *
+	 * @param string $mimetype
+	 *
+	 * @return array|null returns urlsrc and action if matched or null in case of error to retrieve discovery
+	 */
+	public function getWopiSrcUrl(string $mimetype) : ?array {
+		$discoveryXML =	$this->getDiscovery();
+		if ($discoveryXML === null) {
+			// error retrieving discovery
+			return null;
+		}
+
+		$result = $discoveryXML->xpath(\sprintf('/wopi-discovery/net-zone/app[@name=\'%s\']/action', $mimetype));
+		if (($result !== false) && (\count($result) > 0)) {
+			return [
+				'urlsrc' => (string)$result[0]['urlsrc'],
+				'action' => (string)$result[0]['name']
+			];
+		}
+
+		// no matching mimetype found
+		return [ 'urlsrc' => null, 'action' => null ];
+	}
+
+	/**
+	 * Return the content of discovery.xml - either from cache, or download it.
+	 *
+	 * @return SimpleXMLElement|null returns discovery or null in case of discovery cannot be retrieved
+	 */
+	private function getDiscovery() : ?SimpleXMLElement {
+		if ($this->appConfig->testUserSessionEnabled()) {
+			$wopiRemote = $this->appConfig->getAppValue('test_wopi_url');
+			$discoveryKey = 'discovery.xml_test';
+		} else {
+			$wopiRemote = $this->appConfig->getAppValue('wopi_url');
+			$discoveryKey = 'discovery.xml';
+		}
+
+		// Provides access to information about the capabilities of a WOPI client
+		// and the mechanisms for invoking those abilities through URIs.
+		$wopiDiscovery = $wopiRemote . '/hosting/discovery';
+
+		// Read the memcached value (if the memcache is installed)
+		$discovery = $this->cache->get($discoveryKey);
+
+		if ($discovery === null) {
+			$this->logger->debug('Fetching {discoveryKey} as not found in cache', ['app' => 'richdocuments', 'discoveryKey' => $discoveryKey]);
+
+			try {
+				// If we are sending query to built-in CODE server, we avoid using IClient::get() method
+				// because of an encoding issue in guzzle: https://github.com/guzzle/guzzle/issues/1758
+				if (\strpos($wopiDiscovery, 'proxy.php') === false) {
+					$wopiClient = $this->httpClient->newClient();
+					$discovery = $wopiClient->get($wopiDiscovery)->getBody();
+				} else {
+					$discovery = \file_get_contents($wopiDiscovery);
+				}
+			} catch (\Exception $e) {
+				$error_message = $e->getMessage();
+
+				if (\preg_match('/^cURL error ([0-9]*):/', $error_message, $matches)) {
```

Comment:
Any reason we don't send the error message directly to the log file? I'd rather not parse messages

Reply:
Hard to judge, but if someone added it, lets keep this.